aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -166,259 +224,333 @@ module aptos_framework::fungible_asset {
+        };
+        deposit(ZERO_ADDRESS, fungible_asset);
+    }
+
+    /// Create a new wallet object to hold fungible asset for the given user account.
+    fun create_wallet_from_account(
+        user_addr: address,
+        metadata: Object<FungibleAssetMetadata>,
+    ): Object<FungibleAssetWallet> {
+        let user = &create_signer::create_signer(user_addr);
+        let constructor_ref = &object::create_typed_object(user, &DomainSeparator {}, create_wallet_seed(&metadata));
+        let wallet_obj = &object::generate_signer(constructor_ref);
+        move_to(wallet_obj, FungibleAssetWallet {
+            metadata,
+            balance: 0,
+            allow_ungated_transfer: true,
+        });
+        move_to(wallet_obj,
+            FungibleAssetWalletEvents {
+                deposit_events: object::new_event_handle<DepositEvent>(wallet_obj),
+                withdraw_events: object::new_event_handle<WithdrawEvent>(wallet_obj),
+                set_ungated_transfer_events: object::new_event_handle<SetUngatedTransferEvent>(wallet_obj),
+            }
+        );
+
+        object::object_from_constructor_ref<FungibleAssetWallet>(constructor_ref)
+    }
+
+    fun get_wallet<T: key>(user: address, metadata: Object<T>): Object<FungibleAssetWallet> {
+        let wallet_addr = create_wallet_address_from_account(user, metadata);
+        object::address_to_object<FungibleAssetWallet>(wallet_addr)
+    }
+
+    fun create_wallet_seed(metadata: &Object<FungibleAssetMetadata>): vector<u8> {
+        let metadata_obj_address = object::object_address(metadata);
+        bcs::to_bytes(&metadata_obj_address)
+    }
+
+    fun deposit_internal(owner: address, fa: FungibleAsset) acquires FungibleAssetWallet, FungibleAssetWalletEvents {
+        let FungibleAsset { metadata, amount } = fa;
+        if (!wallet_exists(owner, metadata)) {
+            create_wallet_from_account(owner, metadata);
+        };
+        let wallet_addr = create_wallet_address_from_account(owner, metadata);
+        let wallet = borrow_global_mut<FungibleAssetWallet>(wallet_addr);
+        wallet.balance = wallet.balance + amount;
+
+        let events = borrow_global_mut<FungibleAssetWalletEvents>(wallet_addr);
+        event::emit_event(&mut events.deposit_events, DepositEvent { metadata, amount });
     }
 
     /// Extract `amount` of fungible asset from `wallet`.
-    fun extract(
-        wallet: &Object<FungibleAssetWallet>,
+    fun withdraw_internal(
+        wallet_addr: address,
         amount: u64,
-    ): FungibleAsset acquires FungibleAssetWallet {
+    ): FungibleAsset acquires FungibleAssetWallet, FungibleAssetWalletEvents {
         assert!(amount != 0, error::invalid_argument(EAMOUNT_CANNOT_BE_ZERO));
-        let wallet = borrow_fungible_asset_mut(wallet);
-        assert!(wallet.allow_ungated_transfer, error::invalid_argument(EUNGATED_TRANSFER_IS_NOT_ALLOWED));
+        let wallet = borrow_global_mut<FungibleAssetWallet>(wallet_addr);
```

Comment:
Should check existence before this line.

Reply:
I mean `FungibleAssetWallet` may not exist. You want users to see the `Resource not exist` error? it is confusing.